### ISSUE TITLE
Add `ignore: ["inside-block"]` to selector-disallowed-list

### ISSUE THREAD

**Issue by yairEO** [stylelint/stylelint#6294]
## What is the problem you're trying to solve?
I also want to disallow top-level selectors which are starting with `&` **because that's invalid CSS code**

```css
&.foo { }
```

(common to `SASS` precompiler language but is not restricted to it) 

## What solution would you like to see?
either top-level flexible regex rule or an explicit rule for anything starting specifically with `&`

**Comment by ybiquitous** [comment]
I've closed this issue because of not using the issue template. Please provide more details like background, reason, use cases, etc.

**Comment by yairEO** [comment]
@ybiquitous -edited. Please re-open to whoever is in charge here can save potentially millions of developers from a junior-coders catastrophic code-bombing such as described above.

**Comment by ybiquitous** [comment]
@yairEO Thanks for using the template. I understand your motivation.

Your request seems to be satisfied by validators like the [`stylelint-csstree-validator`](https://github.com/csstree/stylelint-validator) plugin, the direct use of [CSSTree](https://github.com/csstree/csstree), or Sass compiler, but is it correct?

[CSSTree](https://csstree.github.io/docs/validator.html):

<img width="713" alt="Screen Shot 2022-09-03 at 18 34 25" src="https://user-images.githubusercontent.com/473530/188264887-1bae2e15-2189-4a25-9865-14c654ee16e8.png">

[Sass](https://www.sassmeister.com/):

<img width="603" alt="Screen Shot 2022-09-03 at 18 35 26" src="https://user-images.githubusercontent.com/473530/188264919-2b4f54e6-f3c5-4778-be9b-3d10b2d564a5.png">

See also [our document about validators](https://stylelint.io/about/linting/#validators).

**Comment by jeddy3** [comment]
This is an interesting one.

According to [the latest draft spec](https://drafts.csswg.org/css-nesting/#nest-selector) `&.foo {}` is a valid CSS:

> When used in the selector of a [nested style rule](https://drafts.csswg.org/css-nesting/#nested-style-rule), the [nesting selector](https://drafts.csswg.org/css-nesting/#nesting-selector) represents the elements matched by the parent rule. _When used in any other context, it represents nothing. (That is, it’s valid, but matches no elements.)_

I think the CSSTree validator is currently flagging it as doesn't parse nesting outright yet, e.g.:

<img width="275" alt="Screenshot 2022-09-03 at 15 14 23" src="https://user-images.githubusercontent.com/808227/188274391-120abf58-5431-4271-8720-a4782ce116c7.png">

Stylelint is a good fit for CSS that is valid, but probably wasn't what the author intended to write. However, we probably don't want to add any new nesting rules just yet as the nesting specification is influx ([with the syntax still to be settled on](https://developer.chrome.com/blog/help-css-nesting/)).

I can see us adding a whole suite of rules for nesting when the specification does stabilise, though. For example, [there is plenty of other](https://drafts.csswg.org/css-nesting/#direct) valid nesting CSS that a person would probably want to limit:

```css
a {
  && {}
}
```

Or enforce a particular notation, e.g. [these two are equivalent](https://drafts.csswg.org/css-nesting/#conditionals):

```css
a {
  @media (orientation: landscape) { color: red; }
}

a {
  @media (orientation: landscape) { & { color: red; } }
}
```

I suggest that anyone who is interested in these [create plugins](https://stylelint.io/developer-guide/plugins/) that we can port over to built-in rules once the specification stabilises. 

In the meantime, we can add an `ignore: ["inside-block"]` option to the [`selector-disallowed-list`](https://stylelint.io/user-guide/rules/list/selector-disallowed-list/) rule ([example of prior art](https://stylelint.io/user-guide/rules/list/at-rule-empty-line-before/#inside-block-1)). A user can then craft a regex to disallow the nesting selector (`&`) anywhere in a selector unless the selector is inside of a block.

It's not a perfect solution as the option will apply to all the items in the array, but it'll probably address @yairEO immediate use case.

**Comment by yairEO** [comment]
The spec will be implemented in new browser, while many of today's projects need to support older browsers, of probably 2+ years old, and in that case, even if the spec says something, it's irrelevant for those mentioned projects since that CSS results a bug, which a rule should prevent.

**Comment by jeddy3** [comment]
@yairEO I've labelled the issue as ready to implement. Please consider [contributing](https://stylelint.io/contributing) if you have time.

There are [steps on how to add a new option](https://stylelint.io/developer-guide/rules#add-an-option-to-a-rule) in the Developer guide.

**Comment by yairEO** [comment]
This is how you can do it in React. I think its better I will add some option to disable it, for an easier way

https://codesandbox.io/s/tagify-1114-disable-the-automatic-filtering-in-whitelist-6ekx7w?file=/src/index.js:982-1042

**Comment by mattmanuel90** [comment]
@yairEO 

Glad I found this. I was about to create this issue. Happy for you to raise the PR. Otherwise, I can do it if you don't mind as we'd like this fix for a different use case.

Ours is to disallow root level classes used in our UI lib, to be used at our application codebase to avoid global styling override issues.

The solution just involves disallowing selectors with a parent.type !== 'root' and it works pretty well given the right regexp.

**Comment by jeddy3** [comment]
> The solution just involves disallowing selectors with a ruleNode.parent.type !== 'root' and it works pretty well if given the right regexp.

That's correct. It'd just be the case of returning early if the `ignore: ["inside-block"]` option is set and the `ruleNode.parent.type !== 'root'` condition is true.

> Glad I found this. I was about to create this issue. Happy for you to raise the PR. Otherwise, I can do it if you don't mind as we'd like this fix for a different use case.

@mattmanuel90 as @yairEO hasn't indicated they're working on it, you're welcome to create a pull request yourself.

**Comment by yairEO** [comment]
I'm a single person, not "they"..  I don't have the time capacity as I already am the author of dozens of open-source projects and have a fulltime job and also need to find the time watching funny cats YouTubes while eating watermelons and thus more busy than Elon Must
